batchflow/sampler.py

```diff
@@ -266,63 +198,174 @@ def truncate(self, high=None, low=None, expr=None, prob=0.5):
+        """
+        _up_size = np.random.binomial(size, self._normed[0])
+        _low_size = size - _up_size
+
+        _up = self.left.sample(size=_up_size)
+        _low = self.right.sample(size=_low_size)
+        _sample = np.concatenate([_up, _low])
+        sample = _sample[np.random.permutation(size)]
+
+        return sample
+
+class AndSampler(Sampler):
+    def __init__(self, left, right, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.left = left
+        self.right = right
+
+    def sample(self, size):
+        """ Sampling procedure of a product of two samplers.
+        """
+        _left = self.left.sample(size)
+        _right = self.right.sample(size)
+        return np.concatenate([_left, _right], axis=1)
+
+class ApplySampler(Sampler):
+    def __init__(self, sampler, transform, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.sampler = sampler
+        self.transform = transform
+
+    def sample(self, size):
+        """ Sampling procedure of a sampler subjugated to a transform.
+        """
+        return self.transform(self.sampler.sample(size))
+
+class TruncateSampler(Sampler):
+    def __init__(self, sampler, high=None, low=None, expr=None, prob=0.5, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.sampler = sampler
+        self.high = high
+        self.low = low
+        self.expr = expr
+        self.prob = prob
+
+    def sample(self, size):
+        """ Sampling method of a sampler subjugated to truncation.
+        """
+        if size == 0:
+            return self.sampler.sample(size=0)
+
+        high, low, expr, prob = self.high, self.low, self.expr, self.prob 
+        # set batch-size
+        expectation = size / prob
+        sigma = np.sqrt(size * (1 - prob) / (prob**2))
+        batch_size = int(expectation + 2 * sigma)
+
+        # sample, filter out, concat
+        ctr = 0
+        cumulated = 0
+        samples = []
+        while cumulated < size:
+            # sample points and compute condition-vector
+            sample = self.sampler.sample(size=batch_size)
+            cond = np.ones(shape=batch_size).astype(np.bool)
+            if low is not None:
+                if expr is not None:
+                    cond &= np.greater_equal(expr(sample).reshape(batch_size, -1), low).all(axis=1)
+                else:
+                    cond &= np.greater_equal(sample, low).all(axis=1)
+
+            if high is not None:
+                if expr is not None:
+                    cond &= np.less_equal(expr(sample).reshape(batch_size, -1), high).all(axis=1)
+                else:
+                    cond &= np.less_equal(sample, high).all(axis=1)
+
+            if high is None and low is None:
+                cond &= expr(sample).all(axis=1)
+
+            # check that truncation-prob is not to small
+            _share = np.sum(cond) / batch_size
+            if _share < SMALL_SHARE and ctr > 0:
```

Comment:
Why would we need this? What is the situation where you actually want to get this error?
I mean, in most cases, you just want your generated points: it can take you 0.01ms or 1ms, if you needed to resample 100 times, but rarely do you care for such small margins. You just want your points.
Can we remove this or, at least, make a `I don't care how much time it takes just sample the points` flag, which is set to `True` by default?

Reply:
If anything, it should be a warning: `Resampling points 1232131 times, so make sure you know what you are doing`